### PR TITLE
cmake: detect linker support

### PR DESCRIPTION
**Pull request by tchaikov**
check the linker support of `--version-script` and `--exclude-libs`
flags. and use them only if they are supported.

llvm-ld does not support `--version-script` at this moment. and it's
reported that old versions of llvm-ld do not support `--exclude-libs`
as well.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: cmake/modules/CephChecks.cmake [hunk 1]
```diff
@@ -124,3 +124,14 @@ int main(int argc, char **argv)
 else(CMAKE_SYSTEM_PROCESSOR STREQUAL CMAKE_HOST_SYSTEM_PROCESSOR)
   message(STATUS "Assuming unaligned access is supported")
 endif(CMAKE_SYSTEM_PROCESSOR STREQUAL CMAKE_HOST_SYSTEM_PROCESSOR)
+
+# should use LINK_OPTIONS instead of LINK_LIBRARIES, if we can use cmake v3.14+
+try_compile(HAVE_LINK_VERSION_SCRIPT
+  ${CMAKE_CURRENT_BINARY_DIR}
+  SOURCES ${CMAKE_CURRENT_LIST_DIR}/CephCheck_link.c
+  LINK_LIBRARIES "-Wl,--version-script=${CMAKE_CURRENT_LIST_DIR}/CephCheck_link.map")
```
Comment (tchaikov) [inline cmake/modules/CephChecks.cmake]: changelog

- s/LINK_OPTIONS/LINK_LIBRARIES/ to be compatible with cmake pre 3.14.

**Comment by tchaikov** [comment]
@wjwithagen thanks for testing the change. seems we need to expose these symbols anyhow. i will try harder tomorrow.

**Comment by wjwithagen** [comment]
@tchaikov 
Spent a lot of time on it to get it working with the most recent llvm-ld, but never succeeded.
So I decided to take the easy way out: use the GNU linker.

Jenkins runs on a 13.0-CURRENT host Clang/LLVM fairly new: 8.0.1:
``` 
# ld.lld -V
LLD 8.0.1 (FreeBSD 366581-1300004)
```

File: cmake/modules/CephCheck_link.c [hunk 2]
```diff
@@ -0,0 +1,2 @@
+int main()
```
Comment (batrick) [inline cmake/modules/CephCheck_link.c]: run trivia:

`int main;`

is the smallest valid C program.

Comment (batrick) [reply cmake/modules/CephCheck_link.c]: `s/run/fun/`

**Pull request merged**